### PR TITLE
Improved NixOS module for Hydra (WIP)

### PR DESCRIPTION
**Pull request by taktoa**
I'm working on making the Hydra configuration more declarative; I think the current state of affairs where you literally have to reverse-engineer the Perl source to figure out what options you can give in `services.hydra.extraConfig` is embarrassing; you'd think that Hydra, of all projects, would be really usable with NixOS.

My ultimate plan is what I call "Fully Declarative Hydra": every aspect of Hydra's state should be specified in the NixOS module except the build and evaluation history.

### CODE REVIEW

File: hydra-module.nix [hunk 1]
```diff
@@ -64,7 +67,7 @@ with rec {
       };
 
       package = mkOption {
-        type = types.path;
+        type = types.package;
```
Comment (taktoa) [inline hydra-module.nix]: The NixOS module documentation recommends `types.package` over `types.path`, but there may be some people taking advantage of the current state of affairs; feel free to let me know if this is the case.

File: hydra-module.nix [hunk 2]
```diff
@@ -2,44 +2,49 @@
...
-
-  serverEnv = env //
-    { HYDRA_TRACKER = cfg.tracker;
-      COLUMNS = "80";
-      PGPASSFILE = "${baseDir}/pgpass-www"; # grrr
-    } // (optionalAttrs cfg.debugServer { DBIC_TRACE = "1"; });
+  env = {
+    NIX_REMOTE = "daemon";
+    # FIXME: why isn't this removed yet? we've been on 17.03 for a while now
+    SSL_CERT_FILE = "/etc/ssl/certs/ca-certificates.crt"; # Remove in 16.03
```
Comment (domenkozar) [inline hydra-module.nix]: Let's remove it.

**Comment by taktoa** [comment]
@domenkozar I have a lot more options added locally; I'll push later today

**Comment by Ericson2314** [comment]
We could do the indentation modernization in a separate PR too just to get that out of the way.

**Comment by taktoa** [comment]
@domenkozar: I just sent up all the stuff I've written; it's messy and probably doesn't evaluate yet

File: hydra-module.nix [hunk 3]
```diff
@@ -1,45 +1,113 @@
...
 
-let
+with rec {
+  inherit (lib) mkIf mkOption;
 
+  forceWHNF = x: builtins.seq x x;
+  forceDeep = x: builtins.deepSeq x x;
+
+  strings = lib.strings // {
+    intercalate = str: list: lib.concatStrings (lib.intersperse str list);
```
Comment (Gabriel439) [inline hydra-module.nix]: `intercalate` already exists in the standard library as `lib.concatMapStringsSep`

File: hydra-module.nix [hunk 4]
```diff
@@ -1,45 +1,113 @@
...
+# ------------------------------------------------------------------------------
+# FIXME: send this stuff upstream to nixpkgs
+# vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
 
-let
+with rec {
+  inherit (lib) mkIf mkOption;
 
+  forceWHNF = x: builtins.seq x x;
+  forceDeep = x: builtins.deepSeq x x;
```
Comment (Gabriel439) [inline hydra-module.nix]: This pull request does not use `forceWHNF` and `forceDeep`

Also `forceWHNF` does not do anything because `seq x x = x`

File: hydra-module.nix [hunk 5]
```diff
@@ -62,15 +130,16 @@ in
       };
 
       package = mkOption {
-        type = types.path;
-        #default = pkgs.hydra;
+        type = types.package;
```
Comment (Gabriel439) [inline hydra-module.nix]: You should probably avoid making unrelated breaking changes when submitting a large PR, because if this change breaks anything we can't easily roll back

File: hydra-module.nix [hunk 6]
```diff
@@ -175,65 +246,733 @@ in
...
           Also, this Hydra instance will serve those downloaded store paths to
           its users with its own signature attached as if it had built them
           itself, so don't enable this feature unless your active binary caches
           are absolute trustworthy.
         '';
       };
-    };
 
-  };
+      googleClientID = mkOption {
```
Comment (Gabriel439) [inline hydra-module.nix]: Is this `googleClientID` a secret or is it safe to store in the configuration?

File: hydra-module.nix [hunk 7]
```diff
@@ -175,65 +246,733 @@ in
...
+        '';
+      };
 
-    users.extraGroups.hydra = { };
+      maxRequests = mkOption {
+        type = types.int;
+        default = 100;
+        example = 1000;
+        description = ''
+          FIXME: lol
```
Comment (Gabriel439) [inline hydra-module.nix]: Did you intend to keep this FIXME? :)

File: hydra-module.nix [hunk 8]
```diff
@@ -1,45 +1,113 @@
...
+      foldr   = lib.lists.fold;
+      foldr1  = foldToFold1 lists.foldr;
+      foldl   = lib.lists.foldl;
+      foldl1  = foldToFold1 lists.foldl;
+      foldl'  = lib.lists.foldl';
+      foldl1' = foldToFold1 lists.foldl';
+    });
+
+  types = lib.types // {
+    oneof = list: (
```
[the remark this answers is not in the document]
Comment (taktoa) [reply hydra-module.nix]: `oneof` is not the same as `enum`, as `enum` cannot take any NixOS option type; it can only take values.

File: hydra-module.nix [hunk 9]
```diff
@@ -1,45 +1,113 @@
...
+          merge = a: b: (
+            if a != null
+            then (if b != null then { v = f a.v b.v; } else a)
+            else (if b != null then b                  else null));
+        };
+        (fold merge null mlist).value));
+    };
+
+    {
+      foldr   = lib.lists.fold;
```
[the remark this answers is not in the document]
Comment (taktoa) [reply hydra-module.nix]: Yeah, I was testing this on a nixpkgs version without `foldr`.

File: hydra-module.nix (hunk 3 above)
Comment (taktoa) [reply hydra-module.nix]: Good to know!

File: hydra-module.nix (hunk 4 above)
Comment (taktoa) [reply hydra-module.nix]: Good point on `forceWHNF`.

File: hydra-module.nix (hunk 6 above)
Comment (taktoa) [reply hydra-module.nix]: I think it is a secret.

File: temporary.nix [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+    '');
+
+  # ----------------------------------------------------------------------------
+
+  # If you authorize with
+  #   `curl -H "Authorization: foo" https://api.github.com`
+  # then you should set `auth = "foo";`.
+  #
+  # There can be multiple <githubstatus> stanzas.
+  makeGithubStatusStanza = (
```
[the remark this answers is not in the document]
Comment (taktoa) [reply temporary.nix]: the intention is to fold the semantics defined by these functions into `hydra-module.nix` and then delete `temporary.nix` altogether

**Comment by taktoa** [comment]
Related issue: https://github.com/NixOS/nix/issues/1556

File: hydra-module.nix (hunk 6 above)
Comment (taktoa) [reply hydra-module.nix]: This is not a bad idea. Will investigate.

**Comment by domenkozar** [comment]
@taktoa sorry I just have a bit of time over weekends, is this ready for review?

**Comment by taktoa** [comment]
Sorry, not yet, I've been busy with other things.

**Comment by dtzWill** [comment]
Gentle but enthusiastic ping! :)

**Comment by ElvishJerricco** [comment]
Seconding that ping!

File: hydra-module.nix (hunk 6 above)
Comment (ElvishJerricco) [reply hydra-module.nix]: +1. I think it's currently impossible to configure Hydra with the current NixOS module without putting e.g. github auth tokens in the nix store.

Comment (Gabriel439) [reply hydra-module.nix]: @ElvishJerricco: I have an example NixOps deploy of Hydra that patches Hydra to fix this.  You can see the deploy here:

https://github.com/dhall-lang/dhall-lang/tree/master/nixops

... and the patch here:

https://github.com/dhall-lang/dhall-lang/blob/master/nixops/hydra.patch

**Comment by taktoa** [comment]
yep, was actually working on this a couple days ago

**Comment by gilligan** [comment]
@taktoa this hasn’t been touched in a while. I think several people had shown interest in this and you apparently invested quite some work. Do you have any intentions to pick this up again at some point?

Maybe you could try to break this down instead into smaller changes that are easier and more feasible to get done and merged?

It had been 2 years but maybe there is still some interest on your side.

**Comment by taktoa** [comment]
Sorry, but I just don't have the cycles these days. I'd be happy to let someone take this work over.
